### ISSUE TITLE
Bert Based Case add operator needs to be fused with a gemm

### ISSUE THREAD

**Issue by causten** [ROCm/AMDMIGraphX#2486]

**Comment by pfultz2** [comment]
In bert base case there is an `add_kernel` not being fused with the gemm:

```
@34 = reshape_lazy[dims={1, 384, 768}](@32) -> float_type, {1, 384, 768}, {294912, 768, 1}, target_id=0
@35 = multibroadcast[out_lens={1, 768, 768},out_dyn_dims={}](@31) -> float_type, {1, 768, 768}, {0, 768, 1}, target_id=0
@36 = gpu::gemm[alpha=1,beta=0,compute_fp32=0,trans_batch=0,solution_idx=0](@34,@35,@33) -> float_type, {1, 384, 768}, {294912, 768, 1}, target_id=0
@37 = load[offset=1179648,end=2359296](@1) -> float_type, {1, 384, 768}, {294912, 768, 1}, target_id=0
@38 = gpu::code_object[code_object=7616,symbol_name=add_kernel,global=147456,local=1024,](@36,@12,@37) -> float_type, {1, 384, 768}, {294912, 768, 1}, target_id=0
```

I also found using `--trim 1132` reduces the problem down in the model that can still reproduce the issue:

```
./bin/driver compile /onnx/bert_base_cased_1.onnx --fill1 input_ids --input-dim @input_ids 1 384 --trim 1132
```

The `add` should be fused in `find_gemm_pointwise`:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/targets/gpu/fuse_ops.cpp#L553

Need to investigate why its not fusing the operator.

**Comment by lakhinderwalia** [comment]
The output is actually okay. Not a bug.

TL;DR: the standalone add_kernel doesn't represent the Add_112 operator following Matmul_111 of `/onnx/bert_base_cased_1.onnx`. It was upon visual inspection expected to be folded. But it represents another Add operation/step.

Details:
The confusion: the add_kernel `@41` in the compiler output below is assumed to represent the Add_112 operator -- which adds a constant bias to the output of Matmul_111 (in the onnx model). But this add_kernel `@41` is actually adding the results of gemm `@38` (i.e. Matmul_111) and add_layernorm_mul `@12`. The step `@12` represents the fusion at Operator Mul_53. This add_kernel therefore can't fuse with just one of its inputs (Matmul_111).
```
module: "main"
@0 = check_context::migraphx::gpu::context -> float_type, {}, {}, target_id=0
@1 = hip::hip_allocate_memory[shape=int8_type, {27648}, {1},id=main:scratch] -> int8_type, {27648}, {1}, target_id=0
@2 = hip::hip_copy_literal[id=main:@literal:56] -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@3 = load[offset=0,end=4](@1) -> float_type, {1, 1}, {1, 1}, target_id=0
input_ids = @param:input_ids -> int64_type, {1, 1}, {1, 1}, target_id=0
@5 = gpu::code_object[code_object=3552,symbol_name=convert_kernel,global=1,local=1024,](input_ids,@3) -> float_type, {1, 1}, {1, 1}, target_id=0
@6 = hip::hip_copy_literal[id=main:@literal:126] -> float_type, {28996, 768}, {768, 1}, target_id=0
@7 = load[offset=3088,end=6160](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@8 = gpu::code_object[code_object=3600,symbol_name=gather_kernel,global=768,local=1024,](@6,@5,@7) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@9 = hip::hip_copy_literal[id=main:@literal:125] -> float_type, {768}, {1}, target_id=0
@10 = load[offset=9216,end=12288](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@11 = multibroadcast[out_lens={1, 1, 768},out_dyn_dims={}](@9) -> float_type, {1, 1, 768}, {0, 0, 1}, target_id=0
@12 = gpu::code_object[code_object=4600,symbol_name=add_layernorm_mul_kernel,global=256,local=256,](@8,@2,@11,@10) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@13 = hip::hip_copy_literal[id=main:@literal:30] -> float_type, {768, 2304}, {2304, 1}, target_id=0
@14 = hip::hip_copy_literal[id=main:@literal:11] -> float_type, {1, 1, 2304}, {2304, 2304, 1}, target_id=0
@15 = load[offset=12288,end=21504](@1) -> float_type, {1, 1, 2304}, {2304, 2304, 1}, target_id=0
@16 = multibroadcast[out_lens={1, 768, 2304},out_dyn_dims={}](@13) -> float_type, {1, 768, 2304}, {0, 2304, 1}, target_id=0
@17 = gpu::gemm[alpha=1,beta=1,compute_fp32=1,trans_batch=0,solution_idx=0](@12,@16,@14,@15) -> float_type, {1, 1, 2304}, {2304, 2304, 1}, target_id=0
@18 = slice[axes={2},starts={768},ends={1536}](@17) -> float_type, {1, 1, 768}, {2304, 2304, 1}, target_id=0
@19 = reshape_lazy[dims={1, 1, 12, 64}](@18) -> float_type, {1, 1, 12, 64}, {768, 768, 64, 1}, target_id=0
@20 = slice[axes={2},starts={0},ends={768}](@17) -> float_type, {1, 1, 768}, {2304, 2304, 1}, target_id=0
@21 = reshape_lazy[dims={1, 1, 12, 64}](@20) -> float_type, {1, 1, 12, 64}, {768, 768, 64, 1}, target_id=0
@22 = load[offset=96,end=144](@1) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@23 = transpose[permutation={0, 2, 3, 1}](@19) -> float_type, {1, 12, 64, 1}, {768, 64, 1, 768}, target_id=0
@24 = transpose[permutation={0, 2, 1, 3}](@21) -> float_type, {1, 12, 1, 64}, {768, 64, 768, 1}, target_id=0
@25 = gpu::gemm[alpha=0.125,beta=0,compute_fp32=1,trans_batch=0,solution_idx=0](@24,@23,@22) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@26 = load[offset=48,end=96](@1) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@27 = gpu::code_object[code_object=4104,symbol_name=softmax_kernel,global=768,local=64,](@25,@26) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@28 = slice[axes={2},starts={1536},ends={2304}](@17) -> float_type, {1, 1, 768}, {2304, 2304, 1}, target_id=0
@29 = reshape_lazy[dims={1, 1, 12, 64}](@28) -> float_type, {1, 1, 12, 64}, {768, 768, 64, 1}, target_id=0
@30 = transpose[permutation={0, 2, 1, 3}](@29) -> float_type, {1, 12, 1, 64}, {768, 64, 768, 1}, target_id=0
@31 = load[offset=21504,end=24576](@1) -> float_type, {1, 12, 1, 64}, {768, 64, 64, 1}, target_id=0
@32 = gpu::gemm[alpha=1,beta=0,compute_fp32=1,trans_batch=0,solution_idx=0](@27,@30,@31) -> float_type, {1, 12, 1, 64}, {768, 64, 64, 1}, target_id=0
@33 = hip::hip_copy_literal[id=main:@literal:98] -> float_type, {768, 768}, {768, 1}, target_id=0
@34 = load[offset=6144,end=9216](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@35 = transpose[permutation={0, 2, 1, 3}](@32) -> float_type, {1, 1, 12, 64}, {768, 64, 64, 1}, target_id=0
@36 = reshape_lazy[dims={1, 1, 768}](@35) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@37 = multibroadcast[out_lens={1, 768, 768},out_dyn_dims={}](@33) -> float_type, {1, 768, 768}, {0, 768, 1}, target_id=0
@38 = gpu::gemm[alpha=1,beta=0,compute_fp32=1,trans_batch=0,solution_idx=0](@36,@37,@34) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@39 = hip::hip_copy_literal[id=main:@literal:44] -> float_type, {768}, {1}, target_id=0
@40 = load[offset=3072,end=6144](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@41 = gpu::code_object[code_object=3520,symbol_name=add_kernel,global=384,local=1024,](@38,@12,@40) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@42 = hip::hip_copy_literal[id=main:@literal:124] -> float_type, {768}, {1}, target_id=0
@43 = multibroadcast[out_lens={1, 1, 768},out_dyn_dims={}](@39) -> float_type, {1, 1, 768}, {0, 0, 1}, target_id=0
@44 = load[offset=0,end=3072](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@45 = multibroadcast[out_lens={1, 1, 768},out_dyn_dims={}](@42) -> float_type, {1, 1, 768}, {0, 0, 1}, target_id=0
@46 = gpu::code_object[code_object=4560,symbol_name=add_layernorm_mul_kernel,global=256,local=256,](@41,@43,@45,@44) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
```

**Comment by lakhinderwalia** [comment]
To illustrate, add_kernel `@41` listed above: it adds results from these two stages:
Adding the results of matmul_111 & a fused mul_53: 

![matmul_111_add_112](https://github.com/ROCm/AMDMIGraphX/assets/139581206/6fdb33af-bf18-4be5-8c2d-2a54d735b545)

![mul_53_add_54](https://github.com/ROCm/AMDMIGraphX/assets/139581206/693e4e81-aa50-4507-8f55-64ee6d6708d7)

**Comment by pfultz2** [comment]
Yes of course the `add` is adding from a previous step and not a constant. I dont see why that makes a difference, we should still be fusing it together. I see no reason why we wouldn't in this case. 

The reasons we would **not** want to fuse the `add`:
* The `C` matrix is non standard shape as rocblas can't handle it
    - If its constant we can handle the non-standard shape as we can do a conversion, but not really applicable here.  
* The gemm is used in multiple times

Looking at the IR, none of these cases are true so there should be a fusion. So we have a bug somewhere as we are not handling this case.

**Comment by lakhinderwalia** [comment]
Giving `find_gemm_pointwise()` a higher matching priority than `find_layernorm_pointwise()` solves this issue -- _hopefully without breaking something else due to this rearrangement_. 

Here are some equivalent results, with `gemm @40` below absorbing a 3rd input tensor, before feeding into fused `layernorm @45`. 
```
module: "main"
@0 = check_context::migraphx::gpu::context -> float_type, {}, {}, target_id=0
@1 = hip::hip_allocate_memory[shape=int8_type, {27648}, {1},id=main:scratch] -> int8_type, {27648}, {1}, target_id=0
@2 = load[offset=0,end=4](@1) -> float_type, {1, 1}, {1, 1}, target_id=0
input_ids = @param:input_ids -> int64_type, {1, 1}, {1, 1}, target_id=0
@4 = gpu::code_object[code_object=3552,symbol_name=convert_kernel,global=1,local=1024,](input_ids,@2) -> float_type, {1, 1}, {1, 1}, target_id=0
@5 = hip::hip_copy_literal[id=main:@literal:54] -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@6 = hip::hip_copy_literal[id=main:@literal:126] -> float_type, {28996, 768}, {768, 1}, target_id=0
@7 = load[offset=3088,end=6160](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@8 = gpu::code_object[code_object=3600,symbol_name=gather_kernel,global=768,local=1024,](@6,@4,@7) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@9 = hip::hip_copy_literal[id=main:@literal:125] -> float_type, {768}, {1}, target_id=0
@10 = multibroadcast[out_lens={1, 1, 768},out_dyn_dims={}](@9) -> float_type, {1, 1, 768}, {0, 0, 1}, target_id=0
@11 = load[offset=9312,end=12384](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@12 = gpu::code_object[code_object=4600,symbol_name=add_layernorm_mul_kernel,global=256,local=256,](@8,@5,@10,@11) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@13 = hip::hip_copy_literal[id=main:@literal:11] -> float_type, {1, 1, 2304}, {2304, 2304, 1}, target_id=0
@14 = hip::hip_copy_literal[id=main:@literal:37] -> float_type, {768, 2304}, {2304, 1}, target_id=0
@15 = load[offset=15456,end=24672](@1) -> float_type, {1, 1, 2304}, {2304, 2304, 1}, target_id=0
@16 = multibroadcast[out_lens={1, 768, 2304},out_dyn_dims={}](@14) -> float_type, {1, 768, 2304}, {0, 2304, 1}, target_id=0
@17 = gpu::gemm[alpha=1,beta=1,compute_fp32=1,trans_batch=0,solution_idx=0](@12,@16,@13,@15) -> float_type, {1, 1, 2304}, {2304, 2304, 1}, target_id=0
@18 = hip::hip_copy_literal[id=main:@literal:41] -> float_type, {768}, {1}, target_id=0
@19 = hip::hip_copy_literal[id=main:@literal:23] -> float_type, {768}, {1}, target_id=0
@20 = slice[axes={2},starts={0},ends={768}](@17) -> float_type, {1, 1, 768}, {2304, 2304, 1}, target_id=0
@21 = reshape_lazy[dims={1, 1, 12, 64}](@20) -> float_type, {1, 1, 12, 64}, {768, 768, 64, 1}, target_id=0
@22 = slice[axes={2},starts={768},ends={1536}](@17) -> float_type, {1, 1, 768}, {2304, 2304, 1}, target_id=0
@23 = reshape_lazy[dims={1, 1, 12, 64}](@22) -> float_type, {1, 1, 12, 64}, {768, 768, 64, 1}, target_id=0
@24 = load[offset=48,end=96](@1) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@25 = transpose[permutation={0, 2, 3, 1}](@23) -> float_type, {1, 12, 64, 1}, {768, 64, 1, 768}, target_id=0
@26 = transpose[permutation={0, 2, 1, 3}](@21) -> float_type, {1, 12, 1, 64}, {768, 64, 768, 1}, target_id=0
@27 = gpu::gemm[alpha=0.125,beta=0,compute_fp32=1,trans_batch=0,solution_idx=0](@26,@25,@24) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@28 = load[offset=0,end=48](@1) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@29 = gpu::code_object[code_object=4104,symbol_name=softmax_kernel,global=768,local=64,](@27,@28) -> float_type, {1, 12, 1, 1}, {12, 1, 1, 1}, target_id=0
@30 = load[offset=6144,end=9216](@1) -> float_type, {1, 12, 1, 64}, {768, 64, 64, 1}, target_id=0
@31 = slice[axes={2},starts={1536},ends={2304}](@17) -> float_type, {1, 1, 768}, {2304, 2304, 1}, target_id=0
@32 = reshape_lazy[dims={1, 1, 12, 64}](@31) -> float_type, {1, 1, 12, 64}, {768, 768, 64, 1}, target_id=0
@33 = transpose[permutation={0, 2, 1, 3}](@32) -> float_type, {1, 12, 1, 64}, {768, 64, 768, 1}, target_id=0
@34 = gpu::gemm[alpha=1,beta=0,compute_fp32=1,trans_batch=0,solution_idx=0](@29,@33,@30) -> float_type, {1, 12, 1, 64}, {768, 64, 64, 1}, target_id=0
@35 = hip::hip_copy_literal[id=main:@literal:98] -> float_type, {768, 768}, {768, 1}, target_id=0
@36 = transpose[permutation={0, 2, 1, 3}](@34) -> float_type, {1, 1, 12, 64}, {768, 64, 64, 1}, target_id=0
@37 = reshape_lazy[dims={1, 1, 768}](@36) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@38 = load[offset=3072,end=6144](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@39 = multibroadcast[out_lens={1, 768, 768},out_dyn_dims={}](@35) -> float_type, {1, 768, 768}, {0, 768, 1}, target_id=0
@40 = gpu::gemm[alpha=1,beta=1,compute_fp32=1,trans_batch=0,solution_idx=0](@37,@39,@12,@38) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@41 = hip::hip_copy_literal[id=main:@literal:124] -> float_type, {768}, {1}, target_id=0
@42 = load[offset=0,end=3072](@1) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@43 = multibroadcast[out_lens={1, 1, 768},out_dyn_dims={}](@19) -> float_type, {1, 1, 768}, {0, 0, 1}, target_id=0
@44 = multibroadcast[out_lens={1, 1, 768},out_dyn_dims={}](@41) -> float_type, {1, 1, 768}, {0, 0, 1}, target_id=0
@45 = gpu::code_object[code_object=4560,symbol_name=add_layernorm_mul_kernel,global=256,local=256,](@40,@43,@44,@42) -> float_type, {1, 1, 768}, {768, 768, 1}, target_id=0
@46 = hip::hip_copy_literal[id=main:@literal:97] -> float_type, {768, 3072}, {3072, 1}, target_id=0
@47 = load[offset=15360,end=27648](@1) -> float_type, {1, 1, 3072}, {3072, 3072, 1}, target_id=0
@48 = multibroadcast[out_lens={1, 768, 3072},out_dyn_dims={}](@46) -> float_type, {1, 768, 3072}, {0, 3072, 1}, target_id=0
@49 = gpu::gemm[alpha=1,beta=0,compute_fp32=1,trans_batch=0,solution_idx=0](@45,@48,@47) -> float_type, {1, 1, 3072}, {3072, 3072, 1}, target_id=0
@50 = hip::hip_copy_literal[id=main:@literal:45] -> float_type, {1, 3072}, {3072, 1}, target_id=0
@51 = hip::hip_copy_literal[id=main:@literal:48] -> float_type, {3072, 768}, {768, 1}, target_id=0
@52 = multibroadcast[out_lens={1, 1, 3072},out_dyn_dims={}](@50) -> float_type, {1, 1, 3072}, {0, 3072, 1}, target_id=0
@53 = load[offset=3072,end=15360](@1) -> float_type, {1, 1, 3072}, {3072, 3072, 1}, target_id=0
@54 = gpu::code_object[code_object=4456,symbol_name=add_mul_erf_add_mul_kernel,global=1536,local=1024,]
```
